scrapy_poet/repository.py

```diff
@@ -0,0 +1,55 @@
+"""The Injection Middleware needs a standard way to build dependencies for
+the Page Inputs used by the request callbacks. That's why we have created a
+repository of ``PageObjectInputProvider`` subclasses.
+
+You could implement different providers in order to acquire data from multiple
+external sources, for example, Splash or Auto Extract API.
+"""
+import typing
+
+from scrapy_poet import autoextract
+from scrapy_poet.providers import (
+    PageObjectInputProvider,
+    ResponseDataProvider,
+)
+# FIXME: refactor _providers / provides / register,  make a nicer API
+
+providers = {}
```

Comment:
I'm thinking that we should maybe have a registry per spider/crawler in addition to the global one. So when you start a Spider, the global providers are copied to the local spider/crawler registry. And we could also provide some hooks to modify the spider/crawler providers. This can be useful for unit testing for example, where you can be mocking the providers to test an Spider.

Another use case could be to be able to generate providers on the fly. But not sure when would that be required. 

It can also be useful to disable some providers if you don't want them available.

Reply:
That's a good idea. I see the potential coming from it. It's just probably out of scope here, would you mind creating an issue/ticket for it?

Reply:
Not sure it is out of the scope given the idea exposed at https://github.com/scrapinghub/scrapy-poet/pull/34#discussion_r466707862 which is basically forget about having a static registry of providers and have it in the spider. How and where to register the providers seems an important question and I would probably tackle it here. But I also understand if this wants to be done in two parts, first merging this PR, and then open a new one to work on spider lifecycle providers.

Reply:
I'd really like to implement it. I do think it's better to manually register the providers, similarly to what we have to do with Django applications. Let's just leave it to the next pull request because we need to agree where this will take place and that could prevent us from merging the other benefits of this pull request.

Reply:
Makes sense to me :+1: